diff --git a/src/video/nacl/SDL_naclvideo.c b/src/video/nacl/SDL_naclvideo.c
index ef1179d..dc4d152 100644
--- a/src/video/nacl/SDL_naclvideo.c
+++ b/src/video/nacl/SDL_naclvideo.c
@@ -45,10 +45,18 @@ struct WMcursor {
   // software cursor emulation.
 };
 
+double g_nacl_scale = 1.0;
+
+static inline double min(double x, double y) {
+  return x < y ? x : y;
+}
+
 void NACL_SetScreenResolution(int width, int height) {
   if (current_video) {
     current_video->hidden->ow = width;
     current_video->hidden->oh = height;
+    g_nacl_scale = min((double)width / current_video->hidden->w,
+		       (double)height / current_video->hidden->h);
     SDL_PrivateResize(width, height);
   } else {
     g_nacl_video_width = width;
@@ -285,6 +293,8 @@ static void NACL_UpdateRects(_THIS, int numrects, SDL_Rect *rects) {
   for (p = start + 3; p < end; p += 4)
     *p = 0xFF;
 
+  dd->ppb_graphics2d->SetScale(dd->context2d, g_nacl_scale);
+
   // Flush on the main thread.
   for (i = 0; i < numrects; ++i) {
     SDL_Rect *r = rects+i;
